// @azure/monitor-opentelemetry-exporter/dist-esm/src/types.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
/**
 * Performance Counter OpenTelemetry compliant names.
 * @internal
 */
export var OTelPerformanceCounterNames;
(function (OTelPerformanceCounterNames) {
    OTelPerformanceCounterNames["PRIVATE_BYTES"] = "Private_Bytes";
    OTelPerformanceCounterNames["AVAILABLE_BYTES"] = "Available_Bytes";
    OTelPerformanceCounterNames["PROCESSOR_TIME"] = "Processor_Time";
    OTelPerformanceCounterNames["PROCESS_TIME"] = "Process_Time";
    OTelPerformanceCounterNames["REQUEST_RATE"] = "Request_Rate";
    OTelPerformanceCounterNames["REQUEST_DURATION"] = "Request_Execution_Time";
})(OTelPerformanceCounterNames || (OTelPerformanceCounterNames = {}));
/**
 * Breeze Performance Counter names.
 * @internal
 */
export var BreezePerformanceCounterNames;
(function (BreezePerformanceCounterNames) {
    BreezePerformanceCounterNames["PRIVATE_BYTES"] = "\\Process(??APP_WIN32_PROC??)\\Private Bytes";
    BreezePerformanceCounterNames["AVAILABLE_BYTES"] = "\\Memory\\Available Bytes";
    BreezePerformanceCounterNames["PROCESSOR_TIME"] = "\\Processor(_Total)\\% Processor Time";
    BreezePerformanceCounterNames["PROCESS_TIME"] = "\\Process(??APP_WIN32_PROC??)\\% Processor Time";
    BreezePerformanceCounterNames["REQUEST_RATE"] = "\\ASP.NET Applications(??APP_W3SVC_PROC??)\\Requests/Sec";
    BreezePerformanceCounterNames["REQUEST_DURATION"] = "\\ASP.NET Applications(??APP_W3SVC_PROC??)\\Request Execution Time";
})(BreezePerformanceCounterNames || (BreezePerformanceCounterNames = {}));
